app/server.js

```diff
@@ -1,13 +1,240 @@
+    return response.end(JSON.stringify(products));
+  }
+});
+
+// POST login
+myRouter.post("/login", function(request, response) {
+  if (request.body.username && request.body.password) {
+    
+    let user = users.find((user) => {
+      return (
+        (user.login.username == request.body.username) &&
+        (user.login.password == request.body.password)
+      );
+    });
+  
+    if (user) {
+      response.writeHead(200, { "Content-Type": "application/json" });
+      let currentUserAccessToken = accessTokens.find((accessTokenUser) => {
+        return accessTokenUser.username == user.login.username;
+      });
+  
+      if (currentUserAccessToken) {
+        currentUserAccessToken.lastUpdated = new Date();
+        return response.end(JSON.stringify(currentUserAccessToken.token));
+      } else {
+        let currentUserNewAccessToken = {
+          username: user.login.username,
+          lastUpdated: new Date(),
+          token: uid(16)
+        };
+        accessTokens.push(currentUserNewAccessToken);
+        return response.end(JSON.stringify(currentUserNewAccessToken));
+      }
+    } else {
+      response.writeHead(401, "Unauthorized Login Username and/or Password");
+      return response.end();
+    }
+  } else {
+    response.writeHead(400, "Invalid Login Username and/or Password");
+    return response.end();
+  }
+});
+
+// function after valid login for access token
+const getCurrentAccessToken = function(request) {
+  const parsedUrl = require("url").parse(request.url, true);
+  if (parsedUrl.query.accessToken) {
+    let currentUserAccessToken = accessTokens.find((accessToken) => {
+      return (
+        accessToken.token == parsedUrl.query.accessToken && 
+        new Date() - accessToken.lastUpdated < TOKEN_VALIDITY_TIMEOUT
+      );
+    });
+
+    if (currentUserAccessToken) {
+      return currentUserAccessToken;
+    } else {
+      return null;
+    }
+  } else {
+    return null;
+  }
+};
+
+// GET login user's cart
+myRouter.get("/me/cart", function(request, response) {
+  let currentUserAccessToken = getCurrentAccessToken(request);
+  
+  if (!currentUserAccessToken) {
+    response.writeHead(401, "Unauthorized Access, Login Required");
+    return response.end();
+  } else {
+    let user = users.find((user) => {
+      return user.login.username == currentUserAccessToken.username;
+    });
+
+    if (!user) {
+      response.writeHead(401, "Unauthorized Access, Login Required");
+      return response.end();
+    } else {
+      response.writeHead(200, { "Content-Type": "application/json" });
+      return response.end(JSON.stringify(user.cart));
+    }
+  }
+});
+
+// POST product add to the login user's cart
+myRouter.post("/me/cart", function(request, response) {
+  let currentUserAccessToken = getCurrentAccessToken(request);
+  
+  if (!currentUserAccessToken) {
+    response.writeHead(401, "Unauthorized Access, Login Required");
```

Comment:
There seems to be a lot of repeated logic that could be made into a helper function